### ISSUE TITLE
Root loader not being called in CatchBoundary

### ISSUE THREAD

**Issue by elkevinwolf** [remix-run/remix#3757]
### What version of Remix are you using?

1.6.4

### Steps to Reproduce

1. Export something from your root loader, and try to read the data returned by that loader in the `CatchBoundary`.

```tsx
import { json, LoaderFunction } from "@remix-run/node"
import { useMatches } from "@remix-run/react"

import { getBaseURL } from "~/utils/seo"

export const loader: LoaderFunction = ({ request }) => {
  return json({ baseUrl: getBaseURL(request) })
}

const Root: React.FC = () => {
  console.log(JSON.stringify(useMatches(), null, 2))
  return <div>Root</div>
}

export const CatchBoundary = () => {
  console.log(JSON.stringify(useMatches(), null, 2))
  return <div>Catch</div>
}

export default Root
```

2. Navigate to a non-existing route.

### Expected Behavior

This should be logged to the console (taken from an existing route).

```json
[
  {
    "id": "root",
    "pathname": "/",
    "params": {},
    "data": {
      "baseUrl": "http://localhost:3000"
    }
  }
]
```

### Actual Behavior

This was logged to the console (taken from a non-existing route).

```json
[
  {
    "id": "root",
    "pathname": "",
    "params": {}
  }
]

```

**Comment by kiliman** [comment]
Depending on if this is a document request or a data request, Remix will work slightly different. If you're navigating to a non-existent route on a document request (full page), Remix throws the new Response(404) *before* it runs any loaders, so `useMatches` will not have the root data (nor any other data in the route).

However if you've already rendered the initial document and making a client-side navigation to a non-existent route, Remix will still throw the 404 response, but this time, `useMatches` still has the original route data, so that is what is returned.

You can verify this behavior by typing in the URL http://localhost:3000/not-found. It should *not* have root data.

Add a link in your route to `<Link to="not-found"/>`. Click on it and you'll see that root data is present, since the current route had already rendered.

I'm pretty sure this is by design. So I would not rely on that data being present in the catch boundary, since that's typically used to indicate some type of error.

**Comment by revelt** [comment]
This bug/feature has been since the beginning; in practice, for example, it means our 404 pages can't show the correct [theme](https://www.npmjs.com/package/remix-themes). It would be nice if loaders were called in catch boundaries.

**Comment by kiliman** [comment]
The only time you won't get root data is if the error/catch boundary is hit on a full document request. My recommendation is to handle this case in the `handleRequest` export of *entry.server*.

```js
export default function handleRequest(
  request: Request,
  responseStatusCode: number,
  responseHeaders: Headers,
  remixContext: EntryContext
) {

  if (responseStatusCode === 404) {
    // fetch theme from (cookie?, host?)
    // render page with catch boundary
    const markup = renderToString(...)
    return new Response("<!DOCTYPE html>" + markup, {
      status: responseStatusCode,
      headers: responseHeaders,
    });
  }
```

**Comment by elkevinwolf** [comment]
> Depending on if this is a document request or a data request, Remix will work slightly different. If you're navigating to a non-existent route on a document request (full page), Remix throws the new Response(404) _before_ it runs any loaders, so `useMatches` will not have the root data (nor any other data in the route).
> 
> However if you've already rendered the initial document and making a client-side navigation to a non-existent route, Remix will still throw the 404 response, but this time, `useMatches` still has the original route data, so that is what is returned.
> 
> You can verify this behavior by typing in the URL http://localhost:3000/not-found. It should _not_ have root data.
> 
> Add a link in your route to `<Link to="not-found"/>`. Click on it and you'll see that root data is present, since the current route had already rendered.
> 
> I'm pretty sure this is by design. So I would not rely on that data being present in the catch boundary, since that's typically used to indicate some type of error.

@kiliman yeah, as I see it, it is like that by design. The funny thing is that the root loader doesn't get called __only__ when a `404` statusCode is thrown, imo, it should be called always, since it is part of the __App Root__, and not specific to any page.

---

> This bug/feature has been since the beginning; in practice, for example, it means our 404 pages can't show the correct [theme](https://www.npmjs.com/package/remix-themes). It would be nice if loaders were called in catch boundaries.

@revelt have you found any workaround? maybe doing something in `entry.server.ts`?

---



> The only time you won't get root data is if the error/catch boundary is hit on a full document request. My recommendation is to handle this case in the `handleRequest` export of _entry.server_.
> 
> ```js
> export default function handleRequest(
>   request: Request,
>   responseStatusCode: number,
>   responseHeaders: Headers,
>   remixContext: EntryContext
> ) {
> 
>   if (responseStatusCode === 404) {
>     // fetch theme from (cookie?, host?)
>     // render page with catch boundary
>     const markup = renderToString(...)
>     return new Response("<!DOCTYPE html>" + markup, {
>       status: responseStatusCode,
>       headers: responseHeaders,
>     });
>   }
> ```

@kiliman how would you recommend me to get the domain and pass it down to be available to put on my metatags? Is there a way to modify that `remixContext` with some data and read it from my route components / catch boundaries / error boundaries?

**Comment by sergiodxa** [comment]
> This bug/feature has been since the beginning; in practice, for example, it means our 404 pages can't show the correct [theme](https://www.npmjs.com/package/remix-themes). It would be nice if loaders were called in catch boundaries.

The problem is that the loader thrown the response that caused the CatchBoundary to render, that means you can't get the full response of the loader because your loader already run and thrown a response earlier.

What you can do is to add the theme or any other data you need to the body of the thrown response.

**Comment by elkevinwolf** [comment]
@sergiodxa how would you recommend me to pass down the domain name to the 404 page? Maybe doing something in `entry.server.tsx`?

**Comment by sergiodxa** [comment]
@elkevinwolf that would be a way, your entry.server could define a context provider with the domain name and then you can access it, and in entry.client you can do the same reading from `window.location` so both have access

Another way is to read the domain name and pass it as the data of the thrown response

```ts
export function loader({ request }: LoaderArgs) {
  let hostname = getHostname(request.url)
  throw json({ hostname }, 404)
}
```

With this, your CatchBoundary will have access to hostname in useCatch

**Comment by elkevinwolf** [comment]
> @elkevinwolf that would be a way, your entry.server could define a context provider with the domain name and then you can access it, and in entry.client you can do the same reading from `window.location` so both have access
> 
> Another way is to read the domain name and pass it as the data of the thrown response
> 
> ```ts
> export function loader({ request }: LoaderArgs) {
>   let hostname = getHostname(request.url)
>   throw json({ hostname }, 404)
> }
> ```
> 
> With this, your CatchBoundary will have access to hostname in useCatch

Good idea of wrapping with a context. Regarding the `loader`, we face the same problem, that it is not readable from the `Root`'s Catch Boundary. The example you put, I assume is from a regular route, not the `Root` component. Actually, if I throw a `404` response from any route, I am still able to read the `Root`'s loader from the `Root`'s Catch Boundary.

**Comment by elkevinwolf** [comment]
Update: I managed to hack around this by adding a [Splat Route](https://remix.run/docs/en/v1/api/conventions#splat-routes) throwing a 404 error status code.

`app/routes/$.tsx`

```tsx
import { notFound } from "remix-utils"

export function loader() {
  throw notFound(null)
}

export default function NotFound() {
  return null
}
```

**Comment by machour** [comment]
Converting this to a discussion as I'm cleaning up the issues.

If this is a request for a new feature, please post a Proposal in the discussions tabs.
